File: domain/src/main/java/org/openmedstack/eventstore/Repository.java
package org.openmedstack.eventstore;

import org.openmedstack.domain.Aggregate;

import java.util.HashMap;
import java.util.concurrent.*;
import java.util.function.Consumer;

public interface Repository {
    <TAggregate extends Aggregate> CompletableFuture<TAggregate> getById(Class<TAggregate> type, String id);

    <TAggregate extends Aggregate> CompletableFuture<TAggregate> getById(Class<TAggregate> type,String id, int version);

    CompletableFuture<Boolean> save(Aggregate aggregate, Consumer<HashMap<String, Object>> updateHeaders) throws DuplicateCommitException;
}

File: modules/domain/src/main/kotlin/org/openmedstack/eventstore/Repository.kt
package org.openmedstack.eventstore

import org.openmedstack.domain.Aggregate
import org.openmedstack.domain.Memento
import java.util.concurrent.CompletableFuture
import java.util.function.Consumer

interface Repository {
    fun <TAggregate : Aggregate, TMemento : Memento> getById(type: Class<TAggregate>, mementoType: Class<Snapshot<TMemento>>, id: String): CompletableFuture<TAggregate>
    fun <TAggregate : Aggregate, TMemento : Memento> getById(type: Class<TAggregate>, mementoType: Class<Snapshot<TMemento>>, id: String, version: Int): CompletableFuture<TAggregate>
    @Throws(DuplicateCommitException::class)
    fun <TAggregate : Aggregate, TMemento : Memento> save(aggregate: TAggregate, updateHeaders: Consumer<HashMap<String, Any>>): CompletableFuture<Boolean>
}
